emails: Rename followup_day1 and followup_day2 email templates

laurynmm opened the pull request:
Follow-up task from #26182, see [this comment](https://github.com/zulip/zulip/pull/26182#discussion_r1254747279) for specific context.

Renames the onboarding email templates to have more descriptive names:
- followup_day1 -> account_registered
- followup_day2 -> onboarding_zulip_topics
- onboarding_zulip_guide (*no change*)

**Notes**:
- Instead of renaming the templates, creates duplicate templates in the first two commits and then deletes the existing templates in the final commit in the series (after the migration).
- Will make comments on the migration commit for questions / thoughts there. 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on zerver/migrations/0465_rename_followup_day_email_templates.py:
```diff
@@ -0,0 +1,59 @@
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    ScheduledEmail = apps.get_model("zerver", "ScheduledEmail")
+    scheduled_onboarding_emails = ScheduledEmail.objects.filter(type=WELCOME)
```
The furthest in the future that a followup_day2 email is scheduled is 3 days 23 hours from the current time, so we could exclude those from this query ...

```python
limit_for_followup_day2_emails = timezone_now() + datetime.timedelta(days=4)
...
scheduled_onboarding_emails = ScheduledEmail.objects.filter(type=WELCOME).exclude(
    scheduled_timestamp__gt=limit_for_followup_day2_emails
)
```

alexmv replied:
I don't think this table is big enough to merit that optimization, and it has the potential to maybe be subtly wrong.

laurynmm replied:
Yeah, I had the same thought.

laurynmm commented on zerver/migrations/0465_rename_followup_day_email_templates.py:
```diff
@@ -0,0 +1,59 @@
+    scheduled_onboarding_emails = ScheduledEmail.objects.filter(type=WELCOME)
+
+    for email in scheduled_onboarding_emails:
+        data = orjson.loads(email.data)
```
The email data with the `"template_prefix"` information is in a text field from the parent class, `AbstractScheduledJob`. As there wasn't a way to query the database for that specific data, I just wrote a Python loop over the potential emails that need to be updated.

alexmv replied:
I think that's a fine option for this case -- but it _is_ possible to do in the database, despite (currently) being a `text` and not a `jsonb` field:

```python3
from django.db.models import F, Func, JSONField, TextField, Value
from django.db.models.functions import Cast

from zerver.models import ScheduledEmail

sched = ScheduledEmail.objects.annotate(as_jsonb=Cast("data", JSONField())).filter(
    type=WELCOME,
    as_jsonb__template_prefix="zerver/emails/followup_day2"
)
sched.update(
    data=Cast(
        Func(
            F("as_jsonb"),
            Value(["template_prefix"]),
            Value("onboarding_zulip_topics", JSONField()),
            function="jsonb_set",
        ),
        TextField(),
    )
)
```

Only very lightly tested, so take with appropriate grain of salt.

laurynmm replied:
Okay. Rewrote the migration based on the above and wrote a basic migration test. And did some manual testing in the db_shell of the dev environment.

laurynmm commented on zerver/migrations/0465_rename_followup_day_email_templates.py:
```diff
@@ -0,0 +1,59 @@
+    ScheduledEmail.objects.bulk_update(scheduled_onboarding_emails, fields=["data"])
+
+
+def revert_followup_day_email_templates_rename(
```
It was simple enough to write the reverse loop, so I went ahead and did that here.

alexmv commented on zerver/migrations/0465_rename_followup_day_email_templates.py:
```diff
@@ -0,0 +1,59 @@
+    for email in scheduled_onboarding_emails:
+        data = orjson.loads(email.data)
+        template = data["template_prefix"]
+        if template == "followup_day1":
```
Shouldn't this be `zerver/emails/followup_day1`?

laurynmm replied:
Yes, indeed it should. I focused on the name of the template that's used in `EMAIL_TYPES` in `zerver/models.py` vs the template prefix that's used in `send_future_email` to create the scheduled email.

laurynmm replied:
Fixed in updated migration.

alexmv commented on zerver/migrations/0465_rename_followup_day_email_templates.py:
```diff
@@ -0,0 +1,59 @@
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+WELCOME = 1
```
Might be worth a comment where this constant would normally be imported from, so it doesn't seem quite so arbitrary.

laurynmm replied:
Added comment in updated migration.

alexmv reviewed (commented):
Thanks, @laurynmm!  I like the bite-size readable commits.

The `followup_day2` vs `zerver/emails/followup_day2` change is the only one here which I think is really blocking.

laurynmm commented:
**Updates**:
- Rewrote migration to update the field by casting text field to jsonb field, based on review feedback above. Added a basic migration test.
- Fixed strings for "template_prefix" and added comment about scheduled message type to migration per review feedback.
- Removing draft/WIP status on pull request.

alexmv reviewed (commented):
This looks great -- and thanks for writing a migration test for this!  Agree that this is the sort of thing that benefits from such.

IMHO, I think it'd be easier to read (and thus easier to check that we're testing the right thing) if we:
1. Use a data structure on the class that defined the data we are going to insert.  This means that we can share the `templates`/`email_data` variables and have confidence that's the dataset we're acting on.
2. Have that structure use a dict, and not a kitchen-sink-list.  This potentially repeats key names, but (a) different templates may have different keys in the structure anyways, and this frees us up to test that, and (b) is much more explicit about the shape of the data.
3. Use a data structure that defines the expected mapping of old -> new name.  This means that we can easily perform (and understand) the desired transformation programmatically in Python, and compare the Python-transformed value to the SQL-transformed one.

But this is all kinda a design nit.  I totally agree that the test as currently written covers the bases -- and being as this is a migration test which we turn over every once in a while, it's not like we'll need to maintain this test long-term.  So this is not a request for a change, merely a "here's how I'd approach this" note. :)

timabbott commented on zerver/migrations/0467_rename_followup_day_email_templates.py:
```diff
@@ -0,0 +1,95 @@
+) -> None:
+    ScheduledEmail = apps.get_model("zerver", "ScheduledEmail")
+
+    account_registered_emails = ScheduledEmail.objects.annotate(
+        as_jsonb=Cast("data", JSONField())
+    ).filter(type=WELCOME, as_jsonb__template_prefix="zerver/emails/followup_day1")
+    account_registered_emails.update(
+        data=Cast(
+            Func(
+                F("as_jsonb"),
+                Value(["template_prefix"]),
+                Value("zerver/emails/account_registered", JSONField()),
+                function="jsonb_set",
+            ),
+            TextField(),
+        )
+    )
```
Yeah, agreed that'd be a nice tweak, though as a migration we'll run once, it's not absolutely critical.

I'm not gonna block merging on this, since it's good to get migrations in to avoid additional future renumbering pain, but it is certainly reasonable to do a quick follow-up PR to tweak thios.

timabbott commented:
Merged, thanks for doing this cleanup @laurynmm!
